File: SIP.py
import numpy as np
from scipy.sparse import csr_matrix, lil_matrix, dok_matrix, find, eye, csc_matrix, identity
from scipy.sparse.linalg import inv,spsolve,splu
import scipy.io as spio
import argparse
import os
import logging
import time
from generate_mf_mor_data import generate_u, generate_udiff 
from utils.tdIntLinBE_new import *
import matplotlib.pyplot as plt
import scipy.sparse as sp

    
def SIPcore3(G, C, E, ports, threshold=2):
    
    n = G.shape[0]
    
    t1 = time.time()

    # floating_nodes = [i for i in range(n) if abs(G[i,i]) < 1e-12]
    # ports = list(set(floating_nodes) | set(ports))
    m = len(ports) # ports + floating nodes

    non_ports = [i for i in range(n) if i not in ports]
    perm = non_ports + ports
    G = G[perm, :][:, perm]
    C = C[perm, :][:, perm]
    C = C + 1e-15 * eye(n, format='csc') #避免奇异
    E = E[perm, :]
    t2 = time.time()
    logging.info("Permutation time: {}".format(t2 - t1))
    
    subG = G
    subC = C
    for i in range(0, n - m):
        # if i >= 20000:
        #     break
        k = n - i
        
        t3 = time.time()

        diag_vals = np.abs(subG[:-m, :-m].diagonal())
        pivot_idx = np.argmax(diag_vals)
        pivot_val = diag_vals[pivot_idx]

        logging.info(f"Processing column {i+1}/{n-m}..., pivot value: {pivot_val}, pivot index: {pivot_idx}")
        
        if pivot_val < threshold:
            logging.info(f"Column {i+1}, pivot value is too small: {pivot_val}")
            logging.info("End of reduction, pivot value is too small.")
            break

        if pivot_idx != 0:
            # 交换 pivot 节点到第 0 个位置
            idx = np.arange(k)
            idx[[0, pivot_idx]] = idx[[pivot_idx, 0]]
            subG = subG[idx, :][:, idx]
            subC = subC[idx, :][:, idx]

        a_ii = subG[0, 0]
            
        b_i = subG[0, 1:k].toarray().flatten()
        # B_i = G[i+1:n, i+1:n]
        
        M_sub = lil_matrix(eye(k))
        # M_sub[0, 0] = -np.sqrt(a_ii)
        M_sub[0,1:] = -b_i / a_ii
        M_sub = csr_matrix(M_sub)
        
        subG = M_sub.T @ subG @ M_sub
        subC = M_sub.T @ subC @ M_sub
        subG = subG[1:, 1:]
        subC = subC[1:, 1:]
        t4 = time.time()
        logging.info(f"Time taken for column {i+1}: {t4 - t3:.4f} seconds")
    
    G_hat = subG
    C_hat = subC
    E_hat = E[-G_hat.shape[0]:, :]
    logging.info(f"Total reduction time: {time.time() - t1:.4f} seconds")
    
    return G_hat.tocsc(), C_hat.tocsc(), E_hat.tocsc()

def is_singular_lu(G):
    try:
        lu = splu(G)  # 如果成功，则 G 非奇异
        return False
    except RuntimeError as e:
        if "Factor is exactly singular" in str(e):
            return True
        raise e

if __name__ == "__main__":
    
    parser = argparse.ArgumentParser(description='SIP')
    parser.add_argument('--circuit', type=int, default=2, help='Circuit number')
    parser.add_argument("--port_num", type=int, default= 2000)
    parser.add_argument("--threshold", type=float, default=2)
    args = parser.parse_args()
    save_path = os.path.join('/home/fillip/桌面/CPM-MOR/Exp_res/SIP/{}t/'.format(args.circuit))
    if not os.path.exists(save_path):
        os.makedirs(save_path)
    logging.basicConfig(level = logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s',
                        handlers=[logging.StreamHandler(),logging.FileHandler(f"{save_path}/SIP.log")])
    logging.info(args)
    data = spio.loadmat("/home/fillip/桌面/CPM-MOR/IBM_transient/ibmpg{}t.mat".format(args.circuit))
    port_num = args.port_num
    threshold = args.threshold
    logging.info("Circuit : {}".format(args.circuit))
    C, G, B = data['E'] * 1e-0, data['A'], data['B']
    Node_size = C.shape[0]
    B = B.tocsc()
    C = C.tocsc()
    G = G.tocsc()

    # print(is_singular_lu(G))

    # B = B[:, 0:0+port_num]
    B = B
    # output matrix
    O = B
    m = 2
    s = 1j * 2 * np.pi * 1e9
    time_start = time.time()
    
    ports = []
    for j in range(B.shape[1]):
        rows, _ = B[:, j].nonzero()
        if len(rows) > 0:
            ports.append(rows[0])
    ports = np.unique(ports)
    
    Gr, Cr, Br = SIPcore3(-G, C, B, ports=list(ports),threshold=threshold)
    sp.save_npz(f'Gr_{args.circuit}.npz', Gr)
    sp.save_npz(f'Cr_{args.circuit}.npz', Cr)
    sp.save_npz(f'Br_{args.circuit}.npz', Br)
    # Gr = sp.load_npz(f'Gr_{args.circuit}.npz')
    # Cr = sp.load_npz(f'Cr_{args.circuit}.npz')
    # Br = sp.load_npz(f'Br_{args.circuit}.npz')
    Br = Br[:, :port_num]
    nr_size = Cr.shape[0]
    logging.info("Original order size: {}".format(Node_size))
    logging.info("Reduced order size: {}".format(nr_size))
    time_end = time.time()
    logging.info("SIP time: {}".format(time_end - time_start))

    t0 = 0
    tf = 1e-09
    dt = 1e-11
    srcType = 'pulse'
    VS = []
    VS = np.array(VS)
    IS, VS = generate_udiff(args.port_num, args.circuit, seed = 0)

    x0 = np.zeros((C.shape[0], 1))
    B = B[:, :port_num]
    xAll, time1, dtAll, uAll = tdIntLinBE_new(t0, tf, dt, C, -G, B, VS, IS, x0, srcType)
    y = O.T@xAll

    xr0 = np.zeros((Cr.shape[0], 1))
    xrAll, time2, dtAll, urAll = tdIntLinBE_new(t0, tf, dt, Cr, Gr, Br, VS, IS, xr0, srcType)
    y_mor = Br.T@xrAll

    yy = y[0,:]
    yy_mor = y_mor[0,:]
    for i in range(1, port_num):
        yy = yy + y[i,:]
        yy_mor = yy_mor + y_mor[i,:]
    
    port = 1
    plt.plot(time1, yy, color='green', linestyle='-.', marker='*', label='GT', markevery = 35, markersize=6, linewidth=1.5)
    plt.plot(time2, yy_mor, color='purple', linestyle='--', marker='*', label='SIP', markevery = 25, markersize=6, linewidth=1.5)
    plt.xlabel("Time (s)", fontsize=12)
    plt.ylabel("Response result (V)", fontsize=12)
    plt.legend(fontsize=12)
    plt.grid()
    plt.title("SIP test", fontsize=14)
    plt.tight_layout()
    plt.savefig(save_path+'SIP_{}t_{}.png'.format(args.circuit,port_num,port), dpi=300)
    pass
    

    # n, m = 4, 2
    # G = np.array([
    #     [2, -1, 0, 0],
    #     [-1, 3, -1, 0],
    #     [0, -1, 2, -1],
    #     [0, 0, -1, 1]
    # ], dtype=float)
    # C = np.eye(n) 
    # E = np.array([[1, 0], [0, 0], [0, 1], [0, 0]])
    # ports = [2, 3]
    
    # G_hat, C_hat, E_hat = SIPcore(G, C, E, ports)
    # print("化简后的 G_hat:\n", G_hat)
    # print("化简后的 C_hat:\n", C_hat)
    # print("化简后的 E_hat:\n", E_hat)

File: SIP.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <chrono>
#include <Eigen/Sparse>
#include <Eigen/Dense>

using namespace Eigen;
using namespace std;
using namespace std::chrono;

struct SparseMatrices {
    SparseMatrix<double> G_hat;
    SparseMatrix<double> C_hat;
    SparseMatrix<double> E_hat;
};

SparseMatrices SIPcore3(const SparseMatrix<double>& G, 
                       const SparseMatrix<double>& C, 
                       const SparseMatrix<double>& E,
                       const vector<int>& ports, 
                       double threshold = 2.0) {
    
    int n = G.rows();
    int m = ports.size();
    
    auto start_time = high_resolution_clock::now();
    
    // Create permutation vector
    vector<bool> is_port(n, false);
    for (int port : ports) {
        if (port < n) {
            is_port[port] = true;
        }
    }
    
    vector<int> non_ports;
    for (int i = 0; i < n; ++i) {
        if (!is_port[i]) {
            non_ports.push_back(i);
        }
    }
    
    vector<int> perm;
    perm.reserve(n);
    perm.insert(perm.end(), non_ports.begin(), non_ports.end());
    perm.insert(perm.end(), ports.begin(), ports.end());
    
    // Permute matrices
    vector<Triplet<double>> triplets_G, triplets_C, triplets_E;
    triplets_G.reserve(G.nonZeros());
    triplets_C.reserve(C.nonZeros());
    triplets_E.reserve(E.nonZeros());
    
    for (int i = 0; i < n; ++i) {
        int new_i = find(perm.begin(), perm.end(), i) - perm.begin();
        
        // Permute G
        for (SparseMatrix<double>::InnerIterator it(G, i); it; ++it) {
            int new_j = find(perm.begin(), perm.end(), it.col()) - perm.begin();
            triplets_G.emplace_back(new_i, new_j, it.value());
        }
        
        // Permute C
        for (SparseMatrix<double>::InnerIterator it(C, i); it; ++it) {
            int new_j = find(perm.begin(), perm.end(), it.col()) - perm.begin();
            triplets_C.emplace_back(new_i, new_j, it.value());
        }
        
        // Permute E
        for (SparseMatrix<double>::InnerIterator it(E, i); it; ++it) {
            triplets_E.emplace_back(new_i, it.col(), it.value());
        }
    }
    
    SparseMatrix<double> G_perm(n, n);
    SparseMatrix<double> C_perm(n, n);
    SparseMatrix<double> E_perm(n, E.cols());
    
    G_perm.setFromTriplets(triplets_G.begin(), triplets_G.end());
    C_perm.setFromTriplets(triplets_C.begin(), triplets_C.end());
    E_perm.setFromTriplets(triplets_E.begin(), triplets_E.end());
    
    // Add small diagonal to avoid singularity
    for (int i = 0; i < n; ++i) {
        C_perm.coeffRef(i, i) += 1e-15;
    }
    
    auto perm_time = high_resolution_clock::now();
    cout << "Permutation time: " 
         << duration_cast<milliseconds>(perm_time - start_time).count() / 1000.0 
         << " seconds" << endl;
    
    // Initialize submatrices
    SparseMatrix<double> subG = G_perm;
    SparseMatrix<double> subC = C_perm;
    
    for (int i = 0; i < n - m; ++i) {
        int k = n - i;
        
        auto col_start = high_resolution_clock::now();
        
        // Find pivot
        double max_val = 0.0;
        int pivot_idx = 0;
        
        for (int j = 0; j < k - m; ++j) {
            double diag_val = abs(subG.coeff(j, j));
            if (diag_val > max_val) {
                max_val = diag_val;
                pivot_idx = j;
            }
        }
        
        cout << "Processing column " << i + 1 << "/" << n - m 
             << "..., pivot value: " << max_val 
             << ", pivot index: " << pivot_idx << endl;
        
        if (max_val < threshold) {
            cout << "Column " << i + 1 
                 << ", pivot value is too small: " << max_val << endl;
            cout << "End of reduction, pivot value is too small." << endl;
            break;
        }
        
        // Swap pivot to position 0 if needed
        if (pivot_idx != 0) {
            vector<int> swap_indices(k);
            for (int j = 0; j < k; ++j) {
                swap_indices[j] = j;
            }
            swap(swap_indices[0], swap_indices[pivot_idx]);
            
            // Create permutation matrix for swapping
            SparseMatrix<double> P(k, k);
            vector<Triplet<double>> P_triplets;
            P_triplets.reserve(k);
            for (int j = 0; j < k; ++j) {
                P_triplets.emplace_back(swap_indices[j], j, 1.0);
            }
            P.setFromTriplets(P_triplets.begin(), P_triplets.end());
            
            subG = P.transpose() * subG * P;
            subC = P.transpose() * subC * P;
        }
        
        double a_ii = subG.coeff(0, 0);
        
        // Extract row 0, columns 1 to k-1
        VectorXd b_i(k - 1);
        for (int j = 1; j < k; ++j) {
            b_i(j - 1) = subG.coeff(0, j);
        }
        
        // Create M matrix
        SparseMatrix<double> M(k, k);
        vector<Triplet<double>> M_triplets;
        M_triplets.reserve(k + k - 1);
        
        // Diagonal entries
        for (int j = 0; j < k; ++j) {
            M_triplets.emplace_back(j, j, 1.0);
        }
        
        // Off-diagonal entries
        for (int j = 1; j < k; ++j) {
            M_triplets.emplace_back(0, j, -b_i(j - 1) / a_ii);
        }
        
        M.setFromTriplets(M_triplets.begin(), M_triplets.end());
        
        // Apply transformation
        subG = M.transpose() * subG * M;
        subC = M.transpose() * subC * M;
        
        // Extract submatrices (remove first row and column)
        SparseMatrix<double> new_subG(k - 1, k - 1);
        SparseMatrix<double> new_subC(k - 1, k - 1);
        
        vector<Triplet<double>> new_G_triplets, new_C_triplets;
        new_G_triplets.reserve(subG.nonZeros());
        new_C_triplets.reserve(subC.nonZeros());
        
        for (int j = 1; j < k; ++j) {
            for (SparseMatrix<double>::InnerIterator it(subG, j); it; ++it) {
                if (it.col() >= 1) {
                    new_G_triplets.emplace_back(j - 1, it.col() - 1, it.value());
                }
            }
            
            for (SparseMatrix<double>::InnerIterator it(subC, j); it; ++it) {
                if (it.col() >= 1) {
                    new_C_triplets.emplace_back(j - 1, it.col() - 1, it.value());
                }
            }
        }
        
        new_subG.setFromTriplets(new_G_triplets.begin(), new_G_triplets.end());
        new_subC.setFromTriplets(new_C_triplets.begin(), new_C_triplets.end());
        
        subG = new_subG;
        subC = new_subC;
        
        auto col_end = high_resolution_clock::now();
        cout << "Time taken for column " << i + 1 << ": " 
             << duration_cast<milliseconds>(col_end - col_start).count() / 1000.0 
             << " seconds" << endl;
    }
    
    // Extract E_hat
    int reduced_size = subG.rows();
    SparseMatrix<double> E_hat(reduced_size, E.cols());
    vector<Triplet<double>> E_triplets;
    E_triplets.reserve(E_perm.nonZeros());
    
    for (int i = n - reduced_size; i < n; ++i) {
        for (SparseMatrix<double>::InnerIterator it(E_perm, i); it; ++it) {
            E_triplets.emplace_back(i - (n - reduced_size), it.col(), it.value());
        }
    }
    E_hat.setFromTriplets(E_triplets.begin(), E_triplets.end());
    
    auto total_time = high_resolution_clock::now();
    cout << "Total reduction time: " 
         << duration_cast<milliseconds>(total_time - start_time).count() / 1000.0 
         << " seconds" << endl;
    
    return {subG, subC, E_hat};
}